File: Darwin/YABACore/State/Creation/Linkmark/LinkmarkCreationStateMachine.swift
//
//  LinkmarkCreationStateMachine.swift
//  YABACore
//

import Foundation
import SwiftUI

@MainActor
public final class LinkmarkCreationStateMachine: YabaBaseObservableState<LinkmarkCreationUIState>, YabaScreenStateMachine {
    private var linkFetchTask: Task<Void, Never>?
    private static let urlDebounceNs: UInt64 = 450_000_000

    public override init(initialState: LinkmarkCreationUIState = LinkmarkCreationUIState()) {
        super.init(initialState: initialState)
    }

    public func send(_ event: LinkmarkCreationEvent) async {
        switch event {
        case let .onInit(linkmarkId, initialUrl, initialFolderId, initialTagIds, uncategorizedFolderCreationRequired):
            apply {
                $0.editingBookmarkId = linkmarkId
                if let u = initialUrl { $0.url = u }
                $0.selectedFolderId = initialFolderId
                $0.selectedTagIds = initialTagIds ?? []
                if linkmarkId == nil {
                    $0.uncategorizedFolderCreationRequired = uncategorizedFolderCreationRequired
                } else {
                    $0.uncategorizedFolderCreationRequired = false
                }
            }
        case .onCyclePreviewAppearance:
            apply {
                switch $0.bookmarkAppearance {
                case .list: $0.bookmarkAppearance = .card
                case .card: $0.bookmarkAppearance = .grid
                case .grid: $0.bookmarkAppearance = .list
                }
            }
        case let .onChangeUrl(u):
            apply { $0.url = u }
            scheduleDebouncedLinkFetch()
        case let .onChangeLabel(l):
            apply { $0.label = l }
        case let .onChangeDescription(d):
            apply { $0.bookmarkDescription = d }
        case let .onSelectFolderId(id):
            apply {
                $0.selectedFolderId = id
                $0.uncategorizedFolderCreationRequired = false
            }
        case let .onSelectTagIds(ids):
            apply { $0.selectedTagIds = ids }
        case .onClearLabel:
            apply { $0.label = "" }
        case .onClearDescription:
            apply { $0.bookmarkDescription = "" }
        case .onApplyFromMetadata:
            withAnimation {
                apply {
                    if let t = $0.metadataTitle, !t.isEmpty { $0.label = t }
                    if let d = $0.metadataDescription, !d.isEmpty { $0.bookmarkDescription = d }
                }
            }
        case .onRefetch:
            await runLinkFetch()
        case let .onConverterSucceeded(result):
            await applyConverterResult(result)
        case let .onConverterFailed(err):
            apply { $0.converterError = err }
        case .onSave:
            await persistFromState()
        case .onTogglePinned:
            apply { $0.isPinned.toggle() }
        case let .create(
            bookmarkId,
            folderId,
            label,
            bookmarkDescription,
            isPinned,
            tagIds,
            url,
            domain,
            videoUrl,
            audioUrl,
            metadataTitle,
            metadataDescription,
            metadataAuthor,
            metadataDate
        ):
            apply { $0.isSaving = true }
            AllBookmarksManager.queueCreateBookmark(
                bookmarkId: bookmarkId,
                folderId: folderId,
                kind: .link,
                label: label,
                bookmarkDescription: bookmarkDescription,
                isPinned: isPinned,
                tagIds: tagIds
            )
            LinkmarkManager.queueCreateOrUpdateLinkDetails(
                bookmarkId: bookmarkId,
                url: url,
                domain: domain,
                videoUrl: videoUrl,
                audioUrl: audioUrl,
                metadataTitle: metadataTitle,
                metadataDescription: metadataDescription,
                metadataAuthor: metadataAuthor,
                metadataDate: metadataDate
            )
            apply { $0.isSaving = false }
        }
    }

    private func scheduleDebouncedLinkFetch() {
        linkFetchTask?.cancel()
        linkFetchTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: Self.urlDebounceNs)
            guard !Task.isCancelled, let self else { return }
            await self.runLinkFetch()
        }
    }

    private func runLinkFetch() async {
        let trimmed = state.url.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, trimmed.contains(".") else { return }
        if trimmed == state.lastFetchedUrl, state.converterError == nil { return }

        apply { $0.isFetchingLinkContent = true; $0.converterError = nil }
        defer { apply { $0.isFetchingLinkContent = false } }
        do {
            let (conv, readable) = try await LinkmarkUnfurlCoordinator.shared.fetchAndConvert(urlString: trimmed)
            await applyFetchResult(converter: conv, readable: readable)
        } catch {
            apply {
                $0.converterError = String(describing: error)
            }
        }
    }

    private func applyConverterResult(_ result: WebConverterResult) async {
        let readable = await ConverterResultProcessor.process(
            documentJson: result.documentJson,
            assets: result.assets
        )
        await applyFetchResult(converter: result, readable: readable)
    }

    private func applyFetchResult(converter: WebConverterResult, readable: ReadableUnfurl) async {
        let imageData = await Unfurler.downloadPreviewImageBytes(urlString: converter.linkMetadata.image)
        let logoData = await Unfurler.downloadPreviewImageBytes(urlString: converter.linkMetadata.logo)
        let meta = converter.linkMetadata
        let urlTrim = state.url.trimmingCharacters(in: .whitespacesAndNewlines)
        apply {
            $0.lastFetchedUrl = urlTrim
            $0.cleanedUrl = meta.cleanedUrl.nilIfEmpty
            $0.metadataTitle = meta.title
            $0.metadataDescription = meta.description
            $0.metadataAuthor = meta.author
            $0.metadataDate = meta.date
            $0.videoUrl = meta.video
            $0.audioUrl = meta.audio
            $0.previewImageData = imageData
            $0.previewIconData = logoData
            $0.pendingReadableUnfurl = readable
            $0.converterError = nil
        }
        ReadableAssetResolver.shared.register(unfurl: readable)
        // Do not persist readable versions on fetch while editing; user saves via Done (upserts latest) or uses detail refresh.
    }

    private func persistFromState() async {
        let folderId = state.selectedFolderId
        let label = state.label.trimmingCharacters(in: .whitespacesAndNewlines)
        let url = state.url.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let folderId, !folderId.isEmpty else {
            apply { $0.lastError = "Folder required" }
            return
        }
        if state.uncategorizedFolderCreationRequired {
            do {
                try await CoreOperationQueue.shared.queueAndAwait(name: "EnsureUncategorizedFolderVisible") { context in
                    try FolderManager.ensureUncategorizedFolderVisibleInContext(context)
                }
                apply { $0.uncategorizedFolderCreationRequired = false }
            } catch {
                apply {
                    $0.lastError = String(describing: error)
                    $0.isSaving = false
                }
                return
            }
        }
        guard !label.isEmpty else {
            apply { $0.lastError = "Label required" }
            return
        }
        guard !url.isEmpty else {
            apply { $0.lastError = "URL required" }
            return
        }
        apply { $0.lastError = nil; $0.isSaving = true }
        let domain = LinkmarkManager.extractDomain(from: url)
        let videoUrl = state.videoUrl
        let audioUrl = state.audioUrl
        let metadataTitle = state.metadataTitle
        let metadataDescription = state.metadataDescription
        let metadataAuthor = state.metadataAuthor
        let metadataDate = state.metadataDate
        if let bid = state.editingBookmarkId {
            AllBookmarksManager.queueUpdateBookmarkMetadata(
                bookmarkId: bid,
                folderId: folderId,
                kind: .link,
                label: label,
                bookmarkDescription: state.bookmarkDescription.nilIfEmpty,
                isPinned: state.isPinned,
                tagIds: state.selectedTagIds
            )
            LinkmarkManager.queueCreateOrUpdateLinkDetails(
                bookmarkId: bid,
                url: url,
                domain: domain,
                videoUrl: videoUrl,
                audioUrl: audioUrl,
                metadataTitle: metadataTitle,
                metadataDescription: metadataDescription,
                metadataAuthor: metadataAuthor,
                metadataDate: metadataDate
            )
            if let unfurl = state.pendingReadableUnfurl {
                ReadableContentManager.queueUpsertLinkReadableUnfurlFromBookmarkEditor(
                    bookmarkId: bid,
                    unfurl: unfurl
                )
                AllBookmarksManager.queueSetBookmarkPreviewAssets(
                    bookmarkId: bid,
                    imageBytes: state.previewImageData,
                    iconBytes: state.previewIconData
                )
            }
        } else {
            let bookmarkId = UUID().uuidString
            AllBookmarksManager.queueCreateBookmark(
                bookmarkId: bookmarkId,
                folderId: folderId,
                kind: .link,
                label: label,
                bookmarkDescription: state.bookmarkDescription.nilIfEmpty,
                isPinned: state.isPinned,
                tagIds: state.selectedTagIds
            )
            LinkmarkManager.queueCreateOrUpdateLinkDetails(
                bookmarkId: bookmarkId,
                url: url,
                domain: domain,
                videoUrl: videoUrl,
                audioUrl: audioUrl,
                metadataTitle: metadataTitle,
                metadataDescription: metadataDescription,
                metadataAuthor: metadataAuthor,
                metadataDate: metadataDate
            )
            if let unfurl = state.pendingReadableUnfurl {
                ReadableContentManager.queueUpsertLinkReadableUnfurlFromBookmarkEditor(
                    bookmarkId: bookmarkId,
                    unfurl: unfurl
                )
                AllBookmarksManager.queueSetBookmarkPreviewAssets(
                    bookmarkId: bookmarkId,
                    imageBytes: state.previewImageData,
                    iconBytes: state.previewIconData
                )
            }
        }
        apply { $0.isSaving = false }
    }
}

File: Compose/YABA/app/src/main/java/dev/subfly/yaba/core/state/creation/linkmark/LinkmarkCreationStateMachine.kt
@file:OptIn(ExperimentalCoroutinesApi::class)

package dev.subfly.yaba.core.state.creation.linkmark

import dev.subfly.yaba.core.common.IdGenerator
import dev.subfly.yaba.core.filesystem.LinkmarkFileManager
import dev.subfly.yaba.core.managers.AllBookmarksManager
import dev.subfly.yaba.core.managers.FolderManager
import dev.subfly.yaba.core.managers.LinkmarkManager
import dev.subfly.yaba.core.managers.ReadableContentManager
import dev.subfly.yaba.core.managers.TagManager
import dev.subfly.yaba.core.model.utils.BookmarkAppearance
import dev.subfly.yaba.core.model.utils.BookmarkKind
import dev.subfly.yaba.core.model.utils.CardImageSizing
import dev.subfly.yaba.core.preferences.SettingsStores
import dev.subfly.yaba.core.state.base.BaseStateMachine
import dev.subfly.yaba.core.toast.ToastIconType
import dev.subfly.yaba.core.toast.ToastManager
import dev.subfly.yaba.core.unfurl.ConverterResultProcessor
import dev.subfly.yaba.core.webview.normalizeBridgeOptionalString
import dev.subfly.yaba.core.unfurl.UnfurlError
import dev.subfly.yaba.core.unfurl.Unfurler
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlin.time.Duration.Companion.milliseconds

private const val DEBOUNCE_DELAY_MS = 750L

@OptIn(FlowPreview::class)
class LinkmarkCreationStateMachine :
    BaseStateMachine<LinkmarkCreationUIState, LinkmarkCreationEvent>(
        initialState = LinkmarkCreationUIState()
    ) {
    private var isInitialized = false
    private val preferencesStore
        get() = SettingsStores.userPreferences

    // Debounce flow for URL changes
    private val urlDebounceFlow = MutableSharedFlow<String>(extraBufferCapacity = 1)
    private var urlDebounceJob: Job? = null
    private var toastMessages: LinkmarkCreationToastMessages? = null

    override fun onEvent(event: LinkmarkCreationEvent) {
        when (event) {
            is LinkmarkCreationEvent.OnInit -> onInit(event)
            LinkmarkCreationEvent.OnCyclePreviewAppearance -> onCyclePreviewAppearance()
            is LinkmarkCreationEvent.OnChangeUrl -> onChangeUrl(event)
            is LinkmarkCreationEvent.OnChangeLabel -> onChangeLabel(event)
            is LinkmarkCreationEvent.OnChangeDescription -> onChangeDescription(event)
            is LinkmarkCreationEvent.OnSelectFolder -> onSelectFolder(event)
            is LinkmarkCreationEvent.OnSelectTags -> onSelectTags(event)
            is LinkmarkCreationEvent.OnSave -> onSave(event)
            LinkmarkCreationEvent.OnTogglePinned -> onTogglePinned()
            LinkmarkCreationEvent.OnClearLabel -> onClearLabel()
            LinkmarkCreationEvent.OnClearDescription -> onClearDescription()
            LinkmarkCreationEvent.OnApplyFromMetadata -> onApplyFromMetadata()
            LinkmarkCreationEvent.OnRefetch -> onRefetch()
            is LinkmarkCreationEvent.OnConverterSucceeded -> onConverterSucceeded(event)
            is LinkmarkCreationEvent.OnConverterFailed -> onConverterFailed(event)
        }
    }

    private fun onInit(event: LinkmarkCreationEvent.OnInit) {
        if (isInitialized) return
        isInitialized = true
        toastMessages = event.toastMessages

        // Start URL debounce listener
        startUrlDebounceListener()

        launch {
            // Load preferences for initial appearance
            val preferences = preferencesStore.get()
            updateState {
                it.copy(
                    bookmarkAppearance = preferences.preferredBookmarkAppearance,
                    cardImageSizing = preferences.preferredCardImageSizing,
                )
            }

            // If editing existing linkmark
            event.linkmarkIdString?.let { linkmarkId ->
                val existing = LinkmarkManager.getLinkmarkDetail(linkmarkId)
                if (existing != null) {
                    updateState {
                        it.copy(
                            url = existing.url,
                            cleanedUrl = existing.url,
                            host = existing.domain,
                            label = existing.label,
                            description = existing.description ?: "",
                            metadataTitle = existing.metadataTitle,
                            metadataDescription = existing.metadataDescription,
                            metadataAuthor = existing.metadataAuthor,
                            metadataDate = existing.metadataDate,
                            videoUrl = existing.videoUrl,
                            audioUrl = existing.audioUrl,
                            selectedFolder = existing.parentFolder,
                            selectedTags = existing.tags,
                            editingLinkmark = existing,
                            lastFetchedUrl = existing.url,
                            isPinned = existing.isPinned,
                        )
                    }

                    val savedImageBytes = LinkmarkFileManager.readLinkImageBytes(linkmarkId)
                    val savedIconBytes = LinkmarkFileManager.readDomainIconBytes(linkmarkId)

                    if (savedImageBytes != null || savedIconBytes != null) {
                        updateState {
                            it.copy(
                                imageData = savedImageBytes,
                                iconData = savedIconBytes,
                            )
                        }
                    }
                }
            }

            // Pre-fill initial URL if provided (e.g., from share extension)
            if (event.initialUrl != null && event.linkmarkIdString == null) {
                updateState { it.copy(url = event.initialUrl) }
                urlDebounceFlow.emit(event.initialUrl)
            }

            // Pre-select folder if provided
            event.initialFolderId?.let { folderId ->
                val folder = FolderManager.getFolder(folderId)
                if (folder != null) {
                    updateState { it.copy(selectedFolder = folder) }
                }
            }

            // Pre-select tags if provided
            event.initialTagIds?.let { tagIds ->
                val tags = tagIds.mapNotNull { tagId -> TagManager.getTag(tagId) }
                if (tags.isNotEmpty()) {
                    updateState { it.copy(selectedTags = tags) }
                }
            }

            // Ensure a folder is selected (use uncategorized if none)
            if (currentState().selectedFolder == null) {
                val temporaryUncategorized = FolderManager.createUncategorizedFolderModel()
                updateState {
                    it.copy(
                        selectedFolder = temporaryUncategorized,
                        uncategorizedFolderCreationRequired = true,
                    )
                }
            }
        }
    }

    private fun startUrlDebounceListener() {
        urlDebounceJob?.cancel()
        urlDebounceJob = launch {
            urlDebounceFlow
                .debounce(DEBOUNCE_DELAY_MS.milliseconds)
                .distinctUntilChanged()
                .collect { urlToFetch -> fetchLinkData(urlToFetch) }
        }
    }

    private suspend fun fetchLinkData(
        urlString: String,
        force: Boolean = false,
    ) {
        val state = currentState()

        if (urlString.isBlank() || (!force && urlString == state.lastFetchedUrl)) {
            return
        }

        updateState {
            it.copy(
                isLoading = true,
                error = null,
            )
        }

        try {
            val preview = Unfurler.unfurl(urlString)

            if (preview != null) {
                val rawHtml = preview.rawHtml
                val needsConverter = rawHtml.isNullOrBlank().not()

                updateState {
                    it.copy(
                        cleanedUrl = preview.url,
                        host = extractDomain(preview.url),
                        converterHtml = if (needsConverter) rawHtml else null,
                        converterBaseUrl = if (needsConverter) preview.url else null,
                        converterError = null,
                        lastFetchedUrl = urlString,
                        isLoading = needsConverter,
                        error = null,
                    )
                }
                showUnfurlToast(
                    message = toastMessages?.unfurlSuccess,
                    iconType = ToastIconType.SUCCESS,
                )
            } else {
                val fallback = urlString.trim()
                updateState {
                    it.copy(
                        cleanedUrl = fallback,
                        host = extractDomain(fallback),
                        lastFetchedUrl = urlString,
                        isLoading = false,
                        error = LinkmarkCreationError.UnableToUnfurl,
                    )
                }
                showUnfurlToast(
                    message = toastMessages?.unableToUnfurl,
                    iconType = ToastIconType.ERROR,
                )
            }
        } catch (e: UnfurlError.CannotCreateUrl) {
            updateState {
                it.copy(
                    isLoading = false,
                    error = LinkmarkCreationError.InvalidUrl(e.raw),
                )
            }
            showUnfurlToast(
                message = toastMessages?.invalidUrl,
                iconType = ToastIconType.ERROR,
            )
        } catch (_: UnfurlError) {
            val fallback = urlString.trim()
            updateState {
                it.copy(
                    cleanedUrl = fallback,
                    host = extractDomain(fallback),
                    lastFetchedUrl = urlString,
                    isLoading = false,
                    error = LinkmarkCreationError.UnableToUnfurl,
                )
            }
            showUnfurlToast(
                message = toastMessages?.unableToUnfurl,
                iconType = ToastIconType.ERROR,
            )
        } catch (_: Exception) {
            val fallback = urlString.trim()
            updateState {
                it.copy(
                    cleanedUrl = fallback,
                    host = extractDomain(fallback),
                    lastFetchedUrl = urlString,
                    isLoading = false,
                    error = LinkmarkCreationError.FetchFailed,
                )
            }
            showUnfurlToast(
                message = toastMessages?.genericUnfurlError,
                iconType = ToastIconType.ERROR,
            )
        }
    }

    private fun extractDomain(url: String): String {
        val withoutProtocol = url.substringAfter("://", url)
        val candidate = withoutProtocol.substringBefore("/")
        return candidate.substringBefore("?").substringBefore("#")
    }

    private fun showUnfurlToast(
        message: String?,
        iconType: ToastIconType,
    ) {
        val toastMessage = message ?: return
        val acceptLabel = toastMessages?.acceptLabel

        ToastManager.show(
            message = toastMessage,
            acceptText = acceptLabel,
            iconType = iconType,
        )
    }

    private fun onCyclePreviewAppearance() {
        val state = currentState()

        val (nextAppearance, nextSizing) =
            when (state.bookmarkAppearance) {
                BookmarkAppearance.LIST -> BookmarkAppearance.CARD to CardImageSizing.SMALL
                BookmarkAppearance.CARD -> {
                    when (state.cardImageSizing) {
                        CardImageSizing.SMALL -> BookmarkAppearance.CARD to CardImageSizing.BIG
                        CardImageSizing.BIG -> BookmarkAppearance.GRID to state.cardImageSizing
                    }
                }

                BookmarkAppearance.GRID -> BookmarkAppearance.LIST to state.cardImageSizing
            }

        updateState {
            it.copy(
                bookmarkAppearance = nextAppearance,
                cardImageSizing = nextSizing,
            )
        }
    }

    private fun onChangeUrl(event: LinkmarkCreationEvent.OnChangeUrl) {
        if (currentState().isInEditMode) return
        updateState { it.copy(url = event.newUrl) }
        launch { urlDebounceFlow.emit(event.newUrl) }
    }

    private fun onChangeLabel(event: LinkmarkCreationEvent.OnChangeLabel) {
        updateState { it.copy(label = event.newLabel) }
    }

    private fun onChangeDescription(event: LinkmarkCreationEvent.OnChangeDescription) {
        updateState { it.copy(description = event.newDescription) }
    }

    private fun onSelectFolder(event: LinkmarkCreationEvent.OnSelectFolder) {
        updateState {
            it.copy(selectedFolder = event.folder, uncategorizedFolderCreationRequired = false)
        }
    }

    private fun onSelectTags(event: LinkmarkCreationEvent.OnSelectTags) {
        updateState { it.copy(selectedTags = event.tags) }
    }

    private fun onClearLabel() {
        updateState { it.copy(label = "") }
    }

    private fun onClearDescription() {
        updateState { it.copy(description = "") }
    }

    private fun onApplyFromMetadata() {
        val state = currentState()
        val metaTitle = state.metadataTitle?.trim().orEmpty()
        val metaDesc = state.metadataDescription?.trim().orEmpty()
        updateState {
            it.copy(
                label = metaTitle.ifEmpty { it.label },
                description = metaDesc.ifEmpty { it.description },
            )
        }
    }

    private fun onTogglePinned() {
        updateState { it.copy(isPinned = !it.isPinned) }
    }

    private fun onRefetch() {
        if (currentState().isInEditMode) return
        val currentUrl = currentState().url
        if (currentUrl.isNotBlank()) {
            updateState { it.copy(lastFetchedUrl = "") }
            launch {
                fetchLinkData(
                    urlString = currentUrl,
                    force = true,
                )
            }
        }
    }

    private fun onConverterSucceeded(event: LinkmarkCreationEvent.OnConverterSucceeded) {
        if (currentState().isInEditMode) return
        launch {
            val meta = event.linkMetadata
            val imageBytes =
                Unfurler.downloadPreviewImageBytes(meta.image.normalizeBridgeOptionalString())
            val logoBytes = Unfurler.downloadPreviewImageBytes(meta.logo.normalizeBridgeOptionalString())
            val readable = ConverterResultProcessor.process(
                documentJson = event.documentJson,
                assets = event.assets,
            )
            updateState {
                it.copy(
                    cleanedUrl = meta.cleanedUrl,
                    host = extractDomain(meta.cleanedUrl),
                    metadataTitle = meta.title?.normalizeBridgeOptionalString(),
                    metadataDescription = meta.description?.normalizeBridgeOptionalString(),
                    metadataAuthor = meta.author?.normalizeBridgeOptionalString(),
                    metadataDate = meta.date?.normalizeBridgeOptionalString(),
                    videoUrl = meta.video?.normalizeBridgeOptionalString(),
                    audioUrl = meta.audio?.normalizeBridgeOptionalString(),
                    imageData = imageBytes ?: it.imageData,
                    iconData = logoBytes ?: it.iconData,
                    readable = readable,
                    converterHtml = null,
                    converterBaseUrl = null,
                    converterError = null,
                    isLoading = false,
                )
            }
        }
    }

    private fun onConverterFailed(event: LinkmarkCreationEvent.OnConverterFailed) {
        updateState {
            it.copy(
                converterHtml = null,
                converterBaseUrl = null,
                converterError = event.error.message ?: "Conversion failed",
                isLoading = false,
            )
        }
    }

    private fun onSave(event: LinkmarkCreationEvent.OnSave) {
        val state = currentState()
        val title = state.label.trim()
        if (title.isEmpty()) return
        var selectedFolder = state.selectedFolder ?: return

        launch {
            updateState { it.copy(isSaving = true) }

            try {
                if (state.uncategorizedFolderCreationRequired) {
                    selectedFolder = FolderManager.ensureUncategorizedFolderVisible()
                }

                val bookmarkId: String
                if (state.editingLinkmark != null) {
                    bookmarkId = state.editingLinkmark.id

                    AllBookmarksManager.updateBookmarkMetadata(
                        bookmarkId = bookmarkId,
                        folderId = selectedFolder.id,
                        kind = BookmarkKind.LINK,
                        label = title,
                        description = state.description.ifBlank { null },
                        isPinned = state.isPinned,
                        previewImageBytes = null,
                        previewImageExtension = null,
                        previewIconBytes = null,
                    )

                    LinkmarkManager.createOrUpdateLinkDetails(
                        bookmarkId = bookmarkId,
                        url = state.editingLinkmark.url,
                        domain = state.editingLinkmark.domain,
                        videoUrl = state.editingLinkmark.videoUrl,
                        audioUrl = state.editingLinkmark.audioUrl,
                        metadataTitle = state.editingLinkmark.metadataTitle,
                        metadataDescription = state.editingLinkmark.metadataDescription,
                        metadataAuthor = state.editingLinkmark.metadataAuthor,
                        metadataDate = state.editingLinkmark.metadataDate,
                    )

                    // Handle tag changes
                    val existingTagIds = state.editingLinkmark.tags.map { it.id }.toSet()
                    val newTagIds = state.selectedTags.map { it.id }.toSet()

                    // Add new tags
                    state.selectedTags.filter { it.id !in existingTagIds }.forEach { tag ->
                        AllBookmarksManager.addTagToBookmark(tag.id, bookmarkId)
                    }

                    // Remove old tags
                    state.editingLinkmark.tags.filter { it.id !in newTagIds }.forEach { tag ->
                        AllBookmarksManager.removeTagFromBookmark(tag.id, bookmarkId)
                    }
                } else {
                    // Create new base bookmark metadata first (list/grid source of truth)
                    bookmarkId = IdGenerator.newId()
                    AllBookmarksManager.createBookmarkMetadata(
                        id = bookmarkId,
                        folderId = selectedFolder.id,
                        kind = BookmarkKind.LINK,
                        label = title,
                        description = state.description.ifBlank { null },
                        isPinned = state.isPinned,
                        tagIds = state.selectedTags.map { it.id },
                        previewImageBytes = state.imageData,
                        previewImageExtension = "jpeg",
                        previewIconBytes = state.iconData,
                    )

                    // Then create link-specific details (detail screen extras)
                    LinkmarkManager.createOrUpdateLinkDetails(
                        bookmarkId = bookmarkId,
                        url = state.cleanedUrl,
                        domain = state.host,
                        videoUrl = state.videoUrl,
                        audioUrl = state.audioUrl,
                        metadataTitle = state.metadataTitle,
                        metadataDescription = state.metadataDescription,
                        metadataAuthor = state.metadataAuthor,
                        metadataDate = state.metadataDate,
                    )
                }

                // Save readable content if available (immutable versioning)
                state.readable?.let { readable ->
                    ReadableContentManager.saveReadableContent(bookmarkId, readable)
                }

                updateState { it.copy(isSaving = false, error = null) }
                event.onSavedCallback()
            } catch (_: Exception) {
                updateState {
                    it.copy(
                        isSaving = false,
                        error = LinkmarkCreationError.SaveFailed,
                    )
                }
                event.onErrorCallback()
            }
        }
    }

    override fun clear() {
        isInitialized = false
        toastMessages = null
        urlDebounceJob?.cancel()
        urlDebounceJob = null
        super.clear()
    }
}
